### PR TITLE
chore: adopt `nixpkgs` 22.11

### PR DESCRIPTION
**Pull request by ggreif**
This is mostly mechanical, though it appears that `js_of_ocaml` v4.1.0 has a bug (when miscompiling `moc.js`), so we downgrade it to the known-good v4.0.0.

This needs to be followed-up by some differential debugging to figure out why v4.1.0 fails, and then potentially file bug report(s).

Another wrinkle is that the `musl`-version (static) if `python.twisted` doesn't check on Linux, so it has been worked around by disabling checking on (OCaml) `atdgen`.

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -55,7 +55,7 @@ mo-doc: $(SOURCE_ID)
 	@ln -fs $(MO_DOC_TARGET) mo-doc
 
 moc.js: $(SOURCE_ID)
-	dune build $(DUNE_OPTS) $(MOC_JS_TARGET)
+	dune build --verbose $(DUNE_OPTS) $(MOC_JS_TARGET)
```
Comment (ggreif) [inline src/Makefile:58]: undo

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -33,7 +33,7 @@ jobs:
     - run: nix-env -iA nix-build-uncached -f nix/
 
     - name: "nix-build"
-      run: nix-build-uncached --max-jobs 4 -A all-systems-go -build-flags -L
+      run: nix-build-uncached --max-jobs 1 -A tests -build-flags -L
```
Comment (nomeata) [inline .github/workflows/test.yml:36]: Why this change?

Comment (ggreif) [reply .github/workflows/test.yml:36]: so that we don't cross the storage quota of the builder. I have been trying to limit the number of processes (and limiting the build target) so that there is a chance to (slowly) fill up the binary cache and make some progress.

This will be backed out in the final step.

**Comment by ggreif** [comment]
@nomeata Here is the culprit. Building `-A ic-ref-run` in isolation already triggers the OOM killer (or possibly exhausted the storage quota)
```
> [48 of 62] Compiling IC.Canister.Snapshot ( src/IC/Canister/Snapshot.hs, dist/build/IC/Canister/Snapshot.o, dist/build/IC/Canister/Snapshot.dyn_o )
[3139](https://github.com/dfinity/motoko/actions/runs/4078430179/jobs/7028628345#step:7:3140)
       > [49 of 62] Compiling IC.Canister      ( src/IC/Canister.hs, dist/build/IC/Canister.o, dist/build/IC/Canister.dyn_o )
[3140](https://github.com/dfinity/motoko/actions/runs/4078430179/jobs/7028628345#step:7:3141)
       > [50 of 62] Compiling IC.Ref           ( src/IC/Ref.hs, dist/build/IC/Ref.o, dist/build/IC/Ref.dyn_o )
[3141](https://github.com/dfinity/motoko/actions/runs/4078430179/jobs/7028628345#step:7:3142)
       > /nix/store/28hqpbwpzvpff7ldbhxdhzcpdc34lgsa-stdenv-linux/setup: line 1374:   431 Killed                  ./Setup build
[3142](https://github.com/dfinity/motoko/actions/runs/4078430179/jobs/7028628345#step:7:3143)
       For full logs, run 'nix log /nix/store/hkfycpld9551573cc44larmlm75rjihr-ic-hs-0.0.1.drv'.
[3143](https://github.com/dfinity/motoko/actions/runs/4078430179/jobs/7028628345#step:7:3144)
nix-build-uncached: nix build failed: exit status 1
[3144](https://github.com/dfinity/motoko/actions/runs/4078430179/jobs/7028628345#step:7:3145)
```

**Comment by nomeata** [comment]
Hmm, that's annoying. Maybe it makes sense to migrate ic-hs to that nixpkgs release first, to seed the cache - that's how I used to do it, and is also why we are using the same nix cache for both projects.

**Comment by ggreif** [comment]
> Hmm, that's annoying. Maybe it makes sense to migrate ic-hs to that nixpkgs release first, to seed the cache - that's how I used to do it, and is also why we are using the same nix cache for both projects.

@nomeata I opened https://github.com/dfinity/ic-hs/pull/140 for seeding the cache, but that won't build on ubuntu :-( The yak-shaving continues...

**Comment by nomeata** [comment]
Actually, I'm confused: it should import ic-ref-run using the nixpkg specified there, so should always use the cached version built there. So this PR should not break ic-ref-run in any ways. Did you have the cache enabled above when testing?

So either somehow our importing changed and we are somehow changing the imported derivation, or the CI there stopped uploading to the cache, or the cache has lost the file.

**Comment by ggreif** [comment]
My hunch is that the binary cache has evicted that version of the `ic-ref-run` and thus now it has to be re-seeded from a build. But then my environment (build tools for it) also changed due to the bump in the `nixpkgs` change.

**Comment by nomeata** [comment]
Hmm, but wouldn't then now builds on master also start failing? They download ic-ref-run too!

**Comment by nomeata** [comment]
Ok, I think I see what’s happening:

We have this code:
```
let ic-ref-run = ic-hs-pkgs.ic-hs;
  # copy out the binary, to remove dependencies on the libraries
  nixpkgs.runCommandNoCC "ic-ref-run" {} ''
      mkdir -p $out/bin
      cp ${ic-hs-pkgs.ic-hs}/bin/ic-ref-run $out/bin
  ''; in
```

Indeed `ic-hs-pkgs.ic-hs` is the same on both branches, namely
`/nix/store/hkfycpld9551573cc44larmlm75rjihr-ic-hs-0.0.1.drv` / `/nix/store/nk2r2y2pid24z20g2pbps2p6wcv9w9zl-ic-hs-0.0.1`.

But the wrapping gets a different hash:
On master: `/nix/store/xlynyc66zwcmsqbc4x87fihbk4bc1iz0-ic-ref-run.drv` / `/nix/store/1fxr5dfhaxzs86gyvfid7nv2n1jx4972-ic-ref-run`
On your branch: `/nix/store/6wnlmhaf45nlmhx6yxi0vb2308f1l0aa-ic-ref-run.drv` / `/nix/store/kyigbs5nmpgl70xabsdlr5zj8r0m46lx-ic-ref-run`

And indeed, the wrapped one from `master` is on the cache (`curl https://ic-hs-test.cachix.org/1fxr5dfhaxzs86gyvfid7nv2n1jx4972.narinfo`), but the underlying is not.

```
$ curl https://ic-hs-test.cachix.org/nk2r2y2pid24z20g2pbps2p6wcv9w9zl.narinfo 
404 - Not Found
```

This is a bit annoying. Maybe one fix would be to merge https://github.com/dfinity/motoko/pull/3703 to get a newer `ic-hs` which then hopefully is still in the cache?

**Comment by nomeata** [comment]
I’m updating the test output there. The test suite has become quite large and slow…

**Comment by ggreif** [comment]
@nomeata, the gamble worked.

Now the only remaining problem on Ubuntu is that `python.twisted` fails to test. How could I sneak a `doCheck = !stdenv.hostPlatform.isMusl;` in there?

MacOS is fine.

**Comment by nomeata** [comment]
Is it this problem? https://discourse.nixos.org/t/failing-to-build-python3-10-twisted-22-4-0/24769

**Comment by nomeata** [comment]
Why does it try to build it, and not fetch it from the nix cache? I get
```
~/dfinity/motoko $ nix-build -A python3Packages.twisted nix/
these 7 paths will be fetched (4.55 MiB download, 26.58 MiB unpacked):
  /nix/store/04ka9nhxhvvw36m8jhp08l96866zlq2q-python3.10-incremental-22.10.0
  /nix/store/0glgplb0l07k43rdyr66yjyx08c7cnrk-python3.10-zope.event-4.5.0
  /nix/store/acm0mvamczjxb1hz7n65ig8zjgh955gc-python3.10-zope.interface-5.4.0
  /nix/store/avcjil3sl4l7pv60wxpxyplrs0fq516r-python3.10-hyperlink-21.0.0
  /nix/store/hhwz3kx8j888i8icpa2j68bc0gfrln4f-python3.10-automat-20.2.0
  /nix/store/l63025x8a10y5jqjjvkyxk4lazwzn9dd-python3.10-constantly-15.1.0
  /nix/store/wvwj264kq1qhgg2yi5pl7gqd4amr56nn-python3.10-twisted-22.10.0
copying path '/nix/store/l63025x8a10y5jqjjvkyxk4lazwzn9dd-python3.10-constantly-15.1.0' from 'https://cache.nixos.org'...
copying path '/nix/store/04ka9nhxhvvw36m8jhp08l96866zlq2q-python3.10-incremental-22.10.0' from 'https://cache.nixos.org'...
copying path '/nix/store/hhwz3kx8j888i8icpa2j68bc0gfrln4f-python3.10-automat-20.2.0' from 'https://cache.nixos.org'...
copying path '/nix/store/0glgplb0l07k43rdyr66yjyx08c7cnrk-python3.10-zope.event-4.5.0' from 'https://cache.nixos.org'...
copying path '/nix/store/avcjil3sl4l7pv60wxpxyplrs0fq516r-python3.10-hyperlink-21.0.0' from 'https://cache.nixos.org'...
copying path '/nix/store/acm0mvamczjxb1hz7n65ig8zjgh955gc-python3.10-zope.interface-5.4.0' from 'https://cache.nixos.org'...
copying path '/nix/store/wvwj264kq1qhgg2yi5pl7gqd4amr56nn-python3.10-twisted-22.10.0' from 'https://cache.nixos.org'...
/nix/store/wvwj264kq1qhgg2yi5pl7gqd4amr56nn-python3.10-twisted-22.10.0
```
but somehow it’s trying to build something else here.

**Comment by ggreif** [comment]
@nomeata possibly because the `musl` version is not built on Hydra? I'll try to piece the things together...

**Comment by nomeata** [comment]
Ah, right, musl for the static builds. I don't think we need a static twisted :-)

Let me have a look

**Comment by nomeata** [comment]
Ok, let’s try this

File: .github/workflows/test.yml (hunk 2 above)
Comment (ggreif) [reply .github/workflows/test.yml]: It's gone now :-)

**Comment by nomeata** [comment]
All green!

**Comment by ggreif** [comment]
> All green! (But the test suite has become sloooow over the last years)

Thanks for the assistance!

**Review by crusso — APPROVED** [review]
Heroic effort guys. Wish I knew what was going on...
